hashing.go

```diff
@@ -0,0 +1,170 @@
+//go:build go1.18
+// +build go1.18
+
+package xsync
+
+import (
+	"hash/maphash"
+	"reflect"
+	"unsafe"
+)
+
+// makeHashFunc creates a fast hash function for the given comparable type.
+// The only limitation is that the type should not contain interfaces inside
+func makeHashFunc[T comparable]() func(maphash.Seed, T) uint64 {
+	var zero T
+	rt := reflect.TypeOf(&zero).Elem() // Elem() avoids uninformative panics when T is interface
+
+	blocks := hGetMemLayout(rt)
+
+	// empty struct case, should never happen irl
+	if len(blocks) == 0 {
+		return func(seed maphash.Seed, v T) uint64 {
+			return maphash.Bytes(seed, nil)
+		}
+	}
+
+	// string or a struct with a single string field
+	if len(blocks) == 1 && !blocks[0].IsString {
+		block := blocks[0]
+		return func(seed maphash.Seed, v T) uint64 {
+			return maphash.Bytes(seed, asBytes(unsafe.Pointer(&v), block.Offset, block.Length))
+		}
+	}
+
+	// data type is a single contiguous region of memory (e.g. struct with integer fields)
+	if len(blocks) == 1 && blocks[0].IsString {
+		block := blocks[0]
+		return func(seed maphash.Seed, v T) uint64 {
+			return maphash.String(seed, asString(unsafe.Pointer(&v), block.Offset))
+		}
+	}
+
+	// complex data type consisting of multiple blocks
+	{
+		return func(seed maphash.Seed, v T) uint64 {
+			var h maphash.Hash
+			h.SetSeed(seed)
+			ptr := unsafe.Pointer(&v)
+
+			for _, b := range blocks {
+				if b.IsString {
+					h.WriteString(asString(ptr, b.Offset))
+				} else {
+					h.Write(asBytes(ptr, b.Offset, b.Length))
+				}
+			}
+			return h.Sum64()
+		}
+	}
+}
+
+// asString interprets memory at p + offset as a string header and returns the string.
+func asString(p unsafe.Pointer, offset uintptr) string {
+	sp := unsafe.Add(p, offset)
+	return *(*string)(sp)
+}
+
+// asBytes returns block of memory at p + offset as a byte slice.
+func asBytes(p unsafe.Pointer, offset uintptr, length uintptr) []byte {
+	sp := unsafe.Add(p, offset)
+
+	// Fast approach: 4.5 ns/op for int hashing.
+	// Dirty, because runtime thinks that returned slice hash 1Tb of capacity
```

Comment:
Are there any downsides due to what's described here?

Reply:
I wrote this comment, because I do not know how GC can react on such a dirty trick. 
Documentation for unsafe package says:

Unlike in C, it is not valid to advance a pointer just beyond the end of its original allocation:

```
// INVALID: end points outside allocated space.
var s thing
end = unsafe.Pointer(uintptr(unsafe.Pointer(&s)) + unsafe.Sizeof(s))

// INVALID: end points outside allocated space.
b := make([]byte, n)
end = unsafe.Pointer(uintptr(unsafe.Pointer(&b[0])) + uintptr(n))
```